### PR TITLE
feat: add redis to nginx-s3-gateway

### PR DESCRIPTION
**Pull request by netanelC**
This pull request enhances the nginx-s3-gateway image by integrating the Redis feature using OpenResty. The changes made in this PR aim to provide a more feature-rich and versatile nginx-s3-gateway image with the ability to leverage Redis functionality.
You can see the helm config here: https://github.com/MapColonies/3d-helm/pull/19

### CODE REVIEW

File: nginx-s3-gateway/Dockerfile [hunk 1]
```diff
@@ -1,16 +1,93 @@
...
+    --with-ld-opt='-Wl,-z,relro -Wl,-z,now -Wl,-rpath,/usr/local/lib,--as-needed -pie' \
+    --add-dynamic-module=/work/ngx_devel_kit-0.3.3 \
+    --add-dynamic-module=/work/lua-nginx-module-0.10.26rc1 \
+    --add-dynamic-module=/work/srcache-nginx-module-0.33 \
+    --add-dynamic-module=/work/set-misc-nginx-module-0.33 \
+    --add-dynamic-module=/work/headers-more-nginx-module-0.36 \
+    && make -j2 \
+    && make install
+
+## ADD-ONS THAT DOES NOT NEED COMPLIATION
```
Comment (CptSchnitz) [inline nginx-s3-gateway/Dockerfile]:
```suggestion
## ADD-ONS THAT DOES NOT NEED COMPILATION
```

File: nginx-s3-gateway/Dockerfile [hunk 2]
```diff
@@ -1,16 +1,93 @@
-FROM ghcr.io/nginxinc/nginx-s3-gateway/nginx-oss-s3-gateway:latest-20230703
+FROM debian as BUILDER
+
+RUN apt update && apt install -y wget build-essential git zlib1g-dev curl libpcre2-dev libssl-dev
+
+WORKDIR /work
+
+## BASE NGINX 
+RUN wget https://openresty.org/download/nginx-1.25.3.tar.gz \
```
Comment (CptSchnitz) [inline nginx-s3-gateway/Dockerfile]: Use [ARG](https://docs.docker.com/engine/reference/builder/#arg) to set the nginx version.

File: nginx-s3-gateway/Dockerfile [hunk 3]
```diff
@@ -1,16 +1,93 @@
-FROM ghcr.io/nginxinc/nginx-s3-gateway/nginx-oss-s3-gateway:latest-20230703
+FROM debian as BUILDER
```
Comment (CptSchnitz) [inline nginx-s3-gateway/Dockerfile]: Specify a tag for debian instead of latest

File: nginx-s3-gateway/Dockerfile [hunk 4]
```diff
@@ -1,16 +1,93 @@
...
-  chgrp -R root /var/cache/nginx /var/log/nginx && \
-  chmod -R g=u /etc/nginx && \
-  chmod -R +x /docker-entrypoint.d && \
-  addgroup nginx root
+RUN chmod +x /docker-entrypoint.d/21-envsubst-on-nginx-conf.sh
 
 EXPOSE 8080
 
-USER nginx
+USER nginx
```
Comment (CptSchnitz) [inline nginx-s3-gateway/Dockerfile]: new line missing

File: nginx-s3-gateway/README.md [hunk 5]
```diff
@@ -13,49 +13,98 @@ Configuring `nginxinc/nginx-s3-gateway` image to act as an authenticating and ca
 * Protecting a S3 bucket with a WAF
 * Serving static assets from a S3 bucket alongside a dynamic application endpoints all in a single RESTful directory structure
 
-## Build
+# NGINX-S3-Gateway with Redis Feature
```
Comment (CptSchnitz) [inline nginx-s3-gateway/README.md]:
```suggestion
# NGINX-S3-Gateway with Lua and Redis Feature
```

File: nginx-s3-gateway/README.md [hunk 6]
```diff
@@ -13,49 +13,98 @@ Configuring `nginxinc/nginx-s3-gateway` image to act as an authenticating and ca
...
-  <prefix>/nginx-s3-gateway:v1.0.0 -d
-```
+2. **Installation of Build Dependencies:**
+   - Essential packages, including `wget`, `build-essential`, and required libraries, are installed to facilitate the subsequent build process.
 
-```
-docker container exec -it nginx-s3-gateway /bin/bash
-```
+3. **Workspace Configuration:**
+   - A working directory named "work" is created to organize and execute the build process.
```
Comment (CptSchnitz) [inline nginx-s3-gateway/README.md]: No need to specify this.
Remove all the stuff that are obvious from just reading the dockerfile.
Stuff like ENV / the nginx build params and what is actually being moved is more important.

File: nginx-s3-gateway/README.md [hunk 7]
```diff
@@ -13,49 +13,98 @@ Configuring `nginxinc/nginx-s3-gateway` image to act as an authenticating and ca
...
+   -e NGINX_WORKER_PROCESSES=4 \
+   -e PROXY_CACHE_MAX_SIZE=10g \
+   -e REDIS_SKIP_CACHE=0 \
+   -e REDIS_HOST=127.0.0.1 \
+   -e REDIS_PORT=6379 \
+   -e REDIS_CACHE_EXPIRATION_TIME=1000 \
+   -e REDIS_CONNECT_TIMEOUT=100 \
+   -e REDIS_SEND_TIMEOUT=100 \
+   -e REDIS_READ_TIMEOUT=100 \
+   -e REDIS_AUTH=bla \
```
Comment (CptSchnitz) [inline nginx-s3-gateway/README.md]: As the PR doesn't include this now its not needed. same with the redis in the name of the image.

**Review by CptSchnitz — CHANGES_REQUESTED** [review]
Great job!
Please change the docs to focus less on redis and still on s3-gateway.
In addition i think adding a working e2e example would be great, we can talk about it.

File: nginx-s3-gateway/example/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      - AWS_SECRET_ACCESS_KEY=s3-password
+      - CORS_ENABLED=true
+      - NGINX_WORKER_PROCESSES=4
+      - REDIS_HOST=redis
+      - REDIS_PORT=6379
+      - REDIS_SKIP_CACHE=0
+      - REDIS_CACHE_EXPIRATION_TIME=1000
+      - REDIS_CONNECT_TIMEOUT=1000
+      - REDIS_SEND_TIMEOUT=1000
+      - REDIS_READ_TIMEOUT=1000
```
Comment (CptSchnitz) [inline nginx-s3-gateway/example/docker-compose.yml]: move env to env-file

File: nginx-s3-gateway/README.md [hunk 9]
```diff
@@ -107,4 +98,28 @@ This Dockerfile provides a solid foundation for NGINX with Redis support. Future
 ## Use Cases
 
 * When Redis is down, requests will be sent directly to S3
-* When a request is taking longer than expected in redis, the request will be sent to S3
\ No newline at end of file
+* When a request is taking longer than expected in redis, the request will be sent to S3
+
+**## Example Deployment with Docker Compose**
```
Comment (CptSchnitz) [inline nginx-s3-gateway/README.md]: 1. this looks dafuk and metumtam
2. move this to a separate readme inside the example folder

**Review by CptSchnitz — CHANGES_REQUESTED** [review]

File: nginx-s3-gateway/README.md (hunk 9 above)
Comment (netanelC) [reply nginx-s3-gateway/README.md]: Done

File: nginx-s3-gateway/Dockerfile [hunk 10]
```diff
@@ -1,15 +1,96 @@
...
+    --with-ld-opt='-Wl,-z,relro -Wl,-z,now -Wl,-rpath,/usr/local/lib,--as-needed -pie' \
+    --add-dynamic-module=/work/ngx_devel_kit-0.3.3 \
+    --add-dynamic-module=/work/lua-nginx-module-0.10.26rc1 \
+    --add-dynamic-module=/work/srcache-nginx-module-0.33 \
+    --add-dynamic-module=/work/set-misc-nginx-module-0.33 \
+    --add-dynamic-module=/work/headers-more-nginx-module-0.36 \
+    && make -j2 \
+    && make install
+
+## ADD-ONS THAT DOES NOT NEED COMPILATION
```
Comment (shimoncohen) [inline nginx-s3-gateway/Dockerfile]: `DOES` -> `DO`.

File: nginx-s3-gateway/README.md [hunk 11]
```diff
@@ -13,49 +13,89 @@ Configuring `nginxinc/nginx-s3-gateway` image to act as an authenticating and ca
...
-```
-docker container exec -it nginx-s3-gateway /bin/bash
-```
+3. **NGINX Compilation with OpenResty Modules:**
+   - The OpenResty NGINX modules (e.g., lua-nginx-module and srcache-nginx-module) are cloned into the workspace.
+   - LuaJIT is cloned and installed, and environment variables (`LUAJIT_LIB` and `LUAJIT_INC`) are set to ensure correct library paths.
+   - NGINX is configured using the `./configure` script with various modules and options, including dynamic modules for OpenResty.
+   - NGINX is then compiled (`make`) and installed (`make install`).
+
+   To determine the appropriate configuration for building NGINX with additional Lua and dynamic modules, I retrieved the configuration command using "nginx -V" within the desired version of NGINX. This ensures compatibility with the chosen NGINX version.
```
Comment (shimoncohen) [inline nginx-s3-gateway/README.md]: `I` -> `we`.

File: nginx-s3-gateway/example/21-envsubst-on-nginx-conf.sh [hunk 12]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+envsubst '${NGINX_WORKER_PROCESSES}' < /etc/nginx/nginx.conf.template > /etc/nginx/nginx.conf
```
Comment (shimoncohen) [inline nginx-s3-gateway/example/21-envsubst-on-nginx-conf.sh]: Please add a line at the end of the file.

File: nginx-s3-gateway/example/docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    command: server /data --console-address ":9001"
+
+  # Optional: Sets a default bucket in minio
+  createbucket:
+    image: minio/mc
+    depends_on:
+      - minio
+    entrypoint: >
+      /bin/sh -c "
+      /usr/bin/mc alias set myminio http://minio:9000 s3-user s3-password;
```
Comment (shimoncohen) [inline nginx-s3-gateway/example/docker-compose.yml]: Share user and password with `environment` values.

File: nginx-s3-gateway/example/redis-get-key.lua [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+local ok, err = red:connect(os.getenv("REDIS_HOST"), os.getenv("REDIS_PORT"))
+if not ok then
+    ngx.log(ngx.ERR, "Failed to connect to redis, error -> ", err)
+    return
+end
+local data = red:get(key)
+if data == ngx.null then
+    return ngx.exit(404)
+end
+ngx.print(data)
```
Comment (shimoncohen) [inline nginx-s3-gateway/example/redis-get-key.lua]: Please add a line at the end of the file.

File: nginx-s3-gateway/example/redis-set-key.lua [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    ngx.log(ngx.ERR, "Failed to connect to redis, error -> ", err)
+    return
+end
+local data, err = red:set(key, value)
+if data == ngx.null then
+    ngx.log(ngx.ERR, "Didn't managed to set key, error -> ", err)
+    return ngx.exit(500)
+end
+red:expire(key, os.getenv("REDIS_CACHE_EXPIRATION_TIME"))
+ngx.print(data)
```
Comment (shimoncohen) [inline nginx-s3-gateway/example/redis-set-key.lua]: Please add a line at the end of the file.

File: nginx-s3-gateway/example/nginx.conf [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+    sendfile        on;
+    #tcp_nopush     on;
+
+    keepalive_timeout  65;
+
+    #gzip  on;
+
+    include /etc/nginx/conf.d/*.conf;
+}
```
Comment (shimoncohen) [inline nginx-s3-gateway/example/nginx.conf]: Please add a line at the end of the file.

File: nginx-s3-gateway/example/default.conf [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+
+    # Provide a hint to the client on 405 errors of the acceptable request methods
+    error_page 405 @error405;
+    location @error405 {
+        add_header Allow "${LIMIT_METHODS_TO_CSV}" always;
+        return 405;
+    }
+
+    include /etc/nginx/conf.d/gateway/s3_server.conf;
+}
```
Comment (shimoncohen) [inline nginx-s3-gateway/example/default.conf:352]: Please add a line at the end of the file.

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: nginx-s3-gateway/README.md (hunk 11 above)
Comment (syncush) [reply nginx-s3-gateway/README.md]: WE approve

File: nginx-s3-gateway/example/21-envsubst-on-nginx-conf.sh (hunk 12 above)
Comment (syncush) [inline nginx-s3-gateway/example/21-envsubst-on-nginx-conf.sh]: I would add parentheses to improve readability.

File: nginx-s3-gateway/example/README.md [hunk 18]
```diff
@@ -0,0 +1,23 @@
+# Example Deployment with Docker Compose**
+
+In the `examples` folder, you'll find a Docker Compose file (`docker-compose.yml`) along with NGINX configuration files (`nginx.conf` and `default.conf`) and redis configuration files (key handling files). This example demonstrates a simplified deployment of NGINX-S3-Gateway with Redis integration.
```
Comment (syncush) [inline nginx-s3-gateway/example/README.md]:
```suggestion
In the `example` folder, you'll find a Docker Compose file (`docker-compose.yml`) along with NGINX configuration files (`nginx.conf` and `default.conf`) and redis configuration files (key handling files). This example demonstrates a simplified deployment of NGINX-S3-Gateway with Redis integration.
```

File: nginx-s3-gateway/example/minioData/bla.txt [hunk 19]
```diff
@@ -0,0 +1 @@
+BLA
```
Comment (syncush) [inline nginx-s3-gateway/example/minioData/bla.txt]: don't call it "bla", "example" would be more proper.

File: nginx-s3-gateway/example/nginx.conf [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+worker_processes ${NGINX_WORKER_PROCESSES};
+
+error_log  /dev/stdout info;
+pid        /tmp/nginx.pid;
+
+load_module modules/ndk_http_module.so;
+load_module modules/ngx_http_set_misc_module.so;
+load_module modules/ngx_http_lua_module.so;
+load_module modules/ngx_http_srcache_filter_module.so;
+load_module modules/ngx_http_headers_more_filter_module.so;
```
Comment (syncush) [inline nginx-s3-gateway/example/nginx.conf:12]: add a comment about those modules needed for redis.

File: nginx-s3-gateway/example/redis-get-key.lua [hunk 21]
```diff
@@ -0,0 +1,19 @@
+local key = assert(ngx.var.request_uri, "no key found")
+local redis = require "resty.redis"
+local red, err = redis:new()
+if not red then
+    ngx.log(ngx.ERR, "Failed to create redis variable, error -> ", err)
+    return
+end
+red:set_timeouts(os.getenv("REDIS_CONNECT_TIMEOUT"), os.getenv("REDIS_SEND_TIMEOUT"), os.getenv("REDIS_READ_TIMEOUT"))
+red:set_keepalive(os.getenv("REDIS_MAX_IDLE_TIMEOUT"), os.getenv("REDIS_POOL_SIZE"))
```
Comment (syncush) [inline nginx-s3-gateway/example/redis-get-key.lua]: extract env values into variables

File: nginx-s3-gateway/example/redis-set-key.lua [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+local key = assert(ngx.var.request_uri, "no key found")
+local redis = require "resty.redis"
+local red, err = redis:new()
+if not red then
+    ngx.log(ngx.ERR, "Failed to create redis variable, error -> ", err)
+    return    
+end
+red:set_keepalive(os.getenv("REDIS_MAX_IDLE_TIMEOUT"), os.getenv("REDIS_POOL_SIZE"))
+red:set_timeouts(os.getenv("REDIS_CONNECT_TIMEOUT"), os.getenv("REDIS_SEND_TIMEOUT"), os.getenv("REDIS_READ_TIMEOUT"))
+local ok, err = red:connect(os.getenv("REDIS_HOST"), os.getenv("REDIS_PORT"))
```
Comment (syncush) [inline nginx-s3-gateway/example/redis-set-key.lua]: extract env

File: nginx-s3-gateway/example/redis-set-key.lua [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+if not ok then
+    ngx.log(ngx.ERR, "Failed to connect to redis, error -> ", err)
+    return
+end
+local data, err = red:set(key, value)
+if data == ngx.null then
+    ngx.log(ngx.ERR, "Didn't managed to set key, error -> ", err)
```
Comment (syncush) [inline nginx-s3-gateway/example/redis-set-key.lua]:
```suggestion
    ngx.log(ngx.ERR, "failed to set key, error -> ", err)
```

**Review by syncush — CHANGES_REQUESTED** [review]

File: nginx-s3-gateway/example/minioData/bla.txt (hunk 19 above)
Comment (shimoncohen) [reply nginx-s3-gateway/example/minioData/bla.txt]: Seconded.

File: nginx-s3-gateway/Dockerfile (hunk 10 above)
Comment (netanelC) [reply nginx-s3-gateway/Dockerfile]: Done

Comment (netanelC) [reply nginx-s3-gateway/Dockerfile]: DO

File: nginx-s3-gateway/README.md (hunk 11 above)
Comment (netanelC) [reply nginx-s3-gateway/README.md]: Done

File: nginx-s3-gateway/example/21-envsubst-on-nginx-conf.sh (hunk 12 above)
Comment (netanelC) [reply nginx-s3-gateway/example/21-envsubst-on-nginx-conf.sh]: Done

File: nginx-s3-gateway/example/redis-get-key.lua (hunk 14 above)
Comment (netanelC) [reply nginx-s3-gateway/example/redis-get-key.lua]: Done

File: nginx-s3-gateway/example/redis-set-key.lua (hunk 15 above)
Comment (netanelC) [reply nginx-s3-gateway/example/redis-set-key.lua]: Done

File: nginx-s3-gateway/example/redis-set-key.lua (hunk 23 above)
Comment (netanelC) [reply nginx-s3-gateway/example/redis-set-key.lua]: Done

File: nginx-s3-gateway/example/redis-get-key.lua (hunk 21 above)
Comment (netanelC) [reply nginx-s3-gateway/example/redis-get-key.lua]: Done

File: nginx-s3-gateway/example/redis-set-key.lua (hunk 22 above)
Comment (netanelC) [reply nginx-s3-gateway/example/redis-set-key.lua]: Done

File: nginx-s3-gateway/example/README.md (hunk 18 above)
Comment (netanelC) [reply nginx-s3-gateway/example/README.md]: Done

File: nginx-s3-gateway/example/nginx.conf (hunk 20 above)
Comment (netanelC) [reply nginx-s3-gateway/example/nginx.conf:12]: Done

File: nginx-s3-gateway/example/minioData/bla.txt (hunk 19 above)
Comment (netanelC) [reply nginx-s3-gateway/example/minioData/bla.txt]: Done

File: nginx-s3-gateway/example/nginx.conf (hunk 16 above)
Comment (netanelC) [reply nginx-s3-gateway/example/nginx.conf]: Done

File: nginx-s3-gateway/example/default.conf (hunk 17 above)
Comment (netanelC) [reply nginx-s3-gateway/example/default.conf:352]: Done

Comment (netanelC) [reply nginx-s3-gateway/example/default.conf:351]: Done

File: nginx-s3-gateway/example/21-envsubst-on-nginx-conf.sh (hunk 12 above)
Comment (netanelC) [reply nginx-s3-gateway/example/21-envsubst-on-nginx-conf.sh]: This is the syntax from nginx itself. Isn't it better to keep it as it is?
See: https://www.baeldung.com/linux/nginx-config-environment-variables

**Review by syncush — APPROVED** [review]

File: nginx-s3-gateway/example/.env [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+REDIS_PORT=6379
+REDIS_SKIP_CACHE=0
+REDIS_CACHE_EXPIRATION_TIME=1000
+REDIS_CONNECT_TIMEOUT=1000
+REDIS_SEND_TIMEOUT=1000
+REDIS_READ_TIMEOUT=1000
+REDIS_MAX_IDLE_TIMEOUT=10000
+REDIS_POOL_SIZE=100
+REDIS_REQUIRE_PASSWORD=true
+REDIS_PASSWORD="redis-password"
```
Comment (shimoncohen) [inline nginx-s3-gateway/example/.env]: Please add a line at the end of the file.

**Review by shimoncohen — APPROVED** [review]

File: nginx-s3-gateway/example/.env (hunk 24 above)
Comment (netanelC) [reply nginx-s3-gateway/example/.env]: Done

File: nginx-s3-gateway/README.md [hunk 25]
```diff
@@ -13,49 +13,89 @@ Configuring `nginxinc/nginx-s3-gateway` image to act as an authenticating and ca
...
+To build the Docker image with the Redis feature:
 
+```bash
+docker build -t nginx-s3-gateway .
 ```
-docker container stop nginx-s3-gateway
+
+## Running the Container
+
+To run a container based on the newly created image:
```
Comment (AnguIar) [inline nginx-s3-gateway/README.md]: `To run a container` should be enough here

**Review by AnguIar — CHANGES_REQUESTED** [review]

File: nginx-s3-gateway/README.md (hunk 25 above)
Comment (netanelC) [reply nginx-s3-gateway/README.md]: Done

**Review by AnguIar — APPROVED** [review]

**Review by syncush — APPROVED** [review]

**Review by CptSchnitz — APPROVED** [review]

**Pull request merged**